// react-record-webcam/dist/RecordWebcam.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.RecordWebcam = void 0;
var react_1 = __importDefault(require("react"));
var Video_1 = __importDefault(require("./Video"));
var Controls_1 = __importDefault(require("./Controls"));
var record_1 = __importDefault(require("./record"));
var types_1 = require("./types");
var utils_1 = require("./utils");
var RECORDER_OPTIONS = Object.freeze({
    type: 'video',
    mimeType: 'video/mp4',
    video: {
        minWidth: 1280,
        minHeight: 720,
        maxWidth: 1920,
        maxHeight: 1080,
        minAspectRatio: 1.77
    }
});
var initialState = Object.freeze({
    isPreview: false,
    isWebcamOn: false,
    isRecording: false,
    status: types_1.CAMERA_STATUS.CLOSED
});
var RecordWebcam = /** @class */ (function (_super) {
    __extends(RecordWebcam, _super);
    function RecordWebcam(props) {
        var _this = _super.call(this, props) || this;
        _this.state = __assign({}, initialState);
        _this.recorderOptions = __assign(__assign({}, RECORDER_OPTIONS), _this.props.options);
        _this.webcamRef = react_1["default"].createRef();
        _this.previewRef = react_1["default"].createRef();
        _this.closeCamera = _this.closeCamera.bind(_this);
        _this.download = _this.download.bind(_this);
        _this.handleOpenCamera = _this.handleOpenCamera.bind(_this);
        _this.handleError = _this.handleError.bind(_this);
        _this.handleCloseCamera = _this.handleCloseCamera.bind(_this);
        _this.handleRetakeRecording = _this.handleRetakeRecording.bind(_this);
        _this.handleStartRecording = _this.handleStartRecording.bind(_this);
        _this.handleStopRecording = _this.handleStopRecording.bind(_this);
        _this.openCamera = _this.openCamera.bind(_this);
        return _this;
    }
    RecordWebcam.prototype.componentDidUpdate = function (prevProps, prevState) {
        if (this.state.status !== prevState.status) {
            if (this.props.getStatus)
                this.props.getStatus(this.state.status);
        }
    };
    RecordWebcam.prototype.openCamera = function () {
        return __awaiter(this, void 0, void 0, function () {
            var recorder;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, record_1["default"](this.recorderOptions)];
                    case 1:
                        recorder = _a.sent();
                        this.recorder = recorder;
                        if (this.webcamRef.current) {
                            this.webcamRef.current.srcObject = recorder.stream;
                        }
                        return [4 /*yield*/, new Promise(function (resolve) { return setTimeout(resolve, 1700); })];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    RecordWebcam.prototype.closeCamera = function () {
        if (this.recorder.stream.id)
            this.recorder.stream.stop();
    };
    RecordWebcam.prototype.handleCloseCamera = function () {
        if (this.previewRef.current) {
            this.previewRef.current.removeAttribute('src');
            this.previewRef.current.load();
        }
        this.setState(__assign({}, initialState));
        this.closeCamera();
    };
    RecordWebcam.prototype.handleOpenCamera = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        this.setState(__assign(__assign({}, initialState), { isWebcamOn: true, status: types_1.CAMERA_STATUS.INIT }));
                        return [4 /*yield*/, this.openCamera()];
                    case 1:
                        _a.sent();
                        this.setState({
                            status: types_1.CAMERA_STATUS.OPEN
                        });
                        return [3 /*break*/, 3];
                    case 2:
                        error_1 = _a.sent();
                        this.handleError(error_1);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    RecordWebcam.prototype.handleStartRecording = function () {
        return __awaiter(this, void 0, void 0, function () {
            var length_1, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, , 6]);
                        return [4 /*yield*/, this.recorder.startRecording()];
                    case 1:
                        _a.sent();
                        this.setState(__assign(__assign({}, initialState), { isRecording: true, isWebcamOn: true, status: types_1.CAMERA_STATUS.RECORDING }));
                        if (!this.props.recordingLength) return [3 /*break*/, 4];
                        length_1 = this.props.recordingLength * 1000;
                        return [4 /*yield*/, new Promise(function (resolve) { return setTimeout(resolve, length_1); })];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, this.handleStopRecording()];
                    case 3:
                        _a.sent();
                        this.closeCamera();
                        _a.label = 4;
                    case 4: return [3 /*break*/, 6];
                    case 5:
                        error_2 = _a.sent();
                        this.handleError(error_2);
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    RecordWebcam.prototype.handleStopRecording = function () {
        return __awaiter(this, void 0, void 0, function () {
            var blob, preview, error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.recorder.stopRecording()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.recorder.getBlob()];
                    case 2:
                        blob = _a.sent();
                        preview = window.URL.createObjectURL(blob);
                        if (this.previewRef.current) {
                            this.previewRef.current.src = preview;
                        }
                        this.closeCamera();
                        this.setState(__assign(__assign({}, initialState), { isWebcamOn: false, isPreview: true, status: types_1.CAMERA_STATUS.PREVIEW }));
                        return [3 /*break*/, 4];
                    case 3:
                        error_3 = _a.sent();
                        this.handleError(error_3);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    RecordWebcam.prototype.handleRetakeRecording = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.handleOpenCamera()];
                    case 1:
                        _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        error_4 = _a.sent();
                        this.handleError(error_4);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    RecordWebcam.prototype.download = function () {
        return __awaiter(this, void 0, void 0, function () {
            var blob, filename, error_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.recorder.getBlob()];
                    case 1:
                        blob = _a.sent();
                        filename = this.props.downloadFileName
                            ? this.props.downloadFileName + ".mp4"
                            : new Date().getTime() + ".mp4";
                        utils_1.saveFile(filename, blob);
                        return [3 /*break*/, 3];
                    case 2:
                        error_5 = _a.sent();
                        this.handleError(error_5);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    RecordWebcam.prototype.handleError = function (error) {
        this.setState(__assign(__assign({}, initialState), { status: types_1.CAMERA_STATUS.ERROR }));
        console.error({ error: error });
    };
    RecordWebcam.prototype.render = function () {
        return (react_1["default"].createElement(react_1["default"].Fragment, null,
            react_1["default"].createElement("style", null, "\n            ." + this.props.cssNamespace + "__wrapper {\n              display: -webkit-box;\n              display: -ms-flexbox;\n              display: flex;\n              -webkit-box-orient: vertical;\n              -webkit-box-direction: normal;\n              -ms-flex-flow: column nowrap;\n                      flex-flow: column nowrap;\n              -webkit-box-pack: justify;\n              -ms-flex-pack: justify;\n                      justify-content: space-between;\n            }\n            ." + this.props.cssNamespace + "__status {\n              margin: 1rem 0;\n            }\n          "),
            react_1["default"].createElement("div", { className: this.props.cssNamespace + "__wrapper" },
                this.props.render &&
                    this.props.render({
                        isWebcamOn: this.state.isWebcamOn,
                        isRecording: this.state.isRecording,
                        isPreview: this.state.isPreview,
                        openCamera: this.handleOpenCamera,
                        closeCamera: this.handleCloseCamera,
                        start: this.handleStartRecording,
                        stop: this.handleStopRecording,
                        retake: this.handleRetakeRecording,
                        download: this.download,
                        status: this.state.status
                    }),
                !this.props.render && (react_1["default"].createElement("div", { className: this.props.cssNamespace + "__status" }, "Status: " + this.state.status)),
                !this.props.render && (react_1["default"].createElement(Controls_1["default"], { cssNamespace: this.props.cssNamespace, openCamera: this.handleOpenCamera, closeCamera: this.handleCloseCamera, start: this.handleStartRecording, stop: this.handleStopRecording, retake: this.handleRetakeRecording, download: this.download, labels: this.props.controlLabels, showOpenCamera: !this.state.isWebcamOn &&
                        !this.state.isRecording &&
                        !this.state.isPreview, showCloseCamera: this.state.isWebcamOn, showStart: this.state.isWebcamOn &&
                        !this.state.isRecording &&
                        !this.state.isPreview, showStop: this.state.isRecording, showRetake: this.state.isPreview, showDownload: this.state.isPreview })),
                react_1["default"].createElement(Video_1["default"], { cssNamespace: this.props.cssNamespace, style: { display: "" + (this.state.isWebcamOn ? 'block' : 'none') }, autoPlay: true, muted: true, loop: true, ref: this.webcamRef }),
                react_1["default"].createElement(Video_1["default"], { cssNamespace: this.props.cssNamespace, style: { display: "" + (this.state.isPreview ? 'block' : 'none') }, autoPlay: true, muted: true, loop: true, ref: this.previewRef }))));
    };
    RecordWebcam.defaultProps = {
        cssNamespace: types_1.NAMESPACES.CSS
    };
    return RecordWebcam;
}(react_1["default"].PureComponent));
exports.RecordWebcam = RecordWebcam;
