// react-record-webcam/dist/types.js
"use strict";
exports.__esModule = true;
exports.NAMESPACES = exports.BUTTON_LABELS = exports.CAMERA_STATUS = void 0;
var CAMERA_STATUS;
(function (CAMERA_STATUS) {
    CAMERA_STATUS["INIT"] = "INIT";
    CAMERA_STATUS["CLOSED"] = "CLOSED";
    CAMERA_STATUS["OPEN"] = "OPEN";
    CAMERA_STATUS["RECORDING"] = "RECORDING";
    CAMERA_STATUS["PREVIEW"] = "PREVIEW";
    CAMERA_STATUS["ERROR"] = "ERROR";
})(CAMERA_STATUS = exports.CAMERA_STATUS || (exports.CAMERA_STATUS = {}));
var BUTTON_LABELS;
(function (BUTTON_LABELS) {
    BUTTON_LABELS["OPEN"] = "Open camera";
    BUTTON_LABELS["CLOSE"] = "Close camera";
    BUTTON_LABELS["START"] = "Start recording";
    BUTTON_LABELS["STOP"] = "Stop recording";
    BUTTON_LABELS["RETAKE"] = "Retake recording";
    BUTTON_LABELS["DOWNLOAD"] = "Download recording";
})(BUTTON_LABELS = exports.BUTTON_LABELS || (exports.BUTTON_LABELS = {}));
var NAMESPACES;
(function (NAMESPACES) {
    NAMESPACES["CSS"] = "react-record-webcam";
})(NAMESPACES = exports.NAMESPACES || (exports.NAMESPACES = {}));
